semmc/tests/semstore/TestArchPropGen.hs

```diff
@@ -0,0 +1,625 @@
+             Right us -> return us
+
+type TestBoundVar sym = BV.BoundVar sym TestGenArch
+
+-- | OperandPair is a helper to contain both of the parameterized
+-- TestBoundVar and Operand values.
+data OperandPair sym x = OpP (TestGenOperand x) (TestBoundVar sym x)
+
+operandName :: ( WI.BoundVar sym ~ WE.ExprBoundVar t) =>
+               OperandPair sym x -> String
+operandName (OpP _ bv) = show $ WE.bvarName $ BV.unBoundVar bv
+
+
+
+genBoundVar_NatArgFoo :: Monad m => MonadIO m =>
+                         WI.IsSymExprBuilder sym =>
+                         sym -> GenT m (OperandPair sym "Foo")
+genBoundVar_NatArgFoo sym =
+  OpP FooArg . BV.BoundVar <$> genBoundNatVar sym Nothing
+
+genBoundVar_BV32ArgBar :: Monad m => MonadIO m =>
+                          WI.IsSymExprBuilder sym =>
+                          sym -> GenT m (OperandPair sym "Bar")
+genBoundVar_BV32ArgBar sym = do
+  OpP BarArg . BV.BoundVar <$> genBoundBV32Var sym Nothing
+
+genBoundVar_BV32ArgBox :: Monad m => MonadIO m =>
+                          WI.IsSymExprBuilder sym =>
+                          sym -> GenT m (OperandPair sym "Box")
+genBoundVar_BV32ArgBox sym = do
+  n <- HG.element [0..3]
+  OpP (BoxArg n) . BV.BoundVar <$> genBoundBV32Var sym Nothing
+
+----------------------------------------------------------------------
+-- What4.Interface.SymExpr generators
+
+genNatSymExpr :: ( Monad m
+                 , MonadIO m
+                 , WI.IsExprBuilder sym
+                 ) =>
+                 sym ->
+                 GenT m (WI.SymExpr sym BaseNatType)
+-- genNatSymExpr = return $ TestLitNat 3
+genNatSymExpr sym = liftIO $ -- WI.natLit sym 3
+                    do x <- WI.natLit sym 3
+                       y <- WI.natLit sym 5
+                       WI.natAdd sym x y
+
+
+genIntSymExpr :: ( MonadIO m
+                 , WI.IsExprBuilder sym
+                 ) =>
+                 sym -> GenT m (WI.SymExpr sym BaseIntegerType)
+genIntSymExpr sym = liftIO $ WI.intLit sym 9
+
+
+genBV32SymExpr :: ( MonadIO m
+                  , WI.IsExprBuilder sym
+                  , WI.IsSymExprBuilder sym
+                  , SA.IsLocation (SA.Location arch)
+                  , SA.Location arch ~ TestLocation
+                  ) =>
+                  sym
+               -> Set.Set (Some (F.Parameter arch sh))
+               -> PL.List (BV.BoundVar sym arch) sh
+               -> MapF.MapF TestLocation (WI.BoundVar sym)
+               -> GenT m (WI.SymExpr sym (BaseBVType 32))
+genBV32SymExpr sym params opvars litvars = do
+  -- get a set of the possible variables that can be used as sources
+  -- in the defs, which are declared already as freshBoundVars.
+  let optional = catMaybes
+                 [
+                   case F.length params of
+                     0 -> Nothing
+                     _ -> Just (paramExprBV32 sym opvars litvars
+                                =<< HG.element (Set.toList params))
```

Comment:
Just trying to understand this - it chooses a parameter and builds a BV32 expr from it (where the parameter may refer to one of the litvars or opvars)?

Reply:
I think all of this looks good - I'm trying to wrap my head around what the pool of operands for generated expressions looks like - does it only include parameters, or also literals? If just parameters, how large do parameter lists in the generated formulas get?

Reply:
That's correct; the case statement + catMaybes is because the Hedgehog choice/element generators require a non-empty list.  If we have other optional elements they could be added here as well.

The parameters for an expression are comprised of a subset of the operands and an arbitrary number of arbitrarily generated literals.  There are currently no FunctionParameter values generated.

The debugPrint definition in TestUtils can be modified to have the tests print (lots of!) output about the generated stuff.  Here's an example of a formula for the OpSurf operator, which has 4 operands, and there were 3 literal variables generated as well for 2 formula definitions:

```
((operands
  ((user__9Vv58rJWt . 'Box)
   (user__L_WC_omh_PtJGOPMl3UDM2fE . 'Bar)
   (user__XMx4t__CwffO_4Uy_ . 'Box)
   (user__4nGak_7JBcgtPO_6xmkE . 'Box)))
  (in
   (user__9Vv58rJWt user__L_WC_omh_PtJGOPMl3UDM2fE user__XMx4t__CwffO_4Uy_ user__4nGak_7JBcgtPO_6xmkE 'Box_0 'Box_1 'Box_3))
  ((defs
   (('Box_0 #xb6a12ea4)
    ('Box_3
     (bvadd
      (bvneg
       (bvashr
        (bvudiv
         (bvudiv
          (bvand
           (bvmul
            (bvadd
             (bvadd #xffd70707 'Box_0)
             (bvneg
              (bvsdiv 'Box_1 user__XMx4t__CwffO_4Uy_)))
            (bvadd
             #xfe32d9d9
             (bvshl user__4nGak_7JBcgtPO_6xmkE #xff6386e6)))
           'Box_1)
          (bvxor
           (bvand
            (bvadd
             (bvand #xfefe5f8e 'Box_1)
             'Box_3)
            'Box_3)
           (bvshl
            (bvsdiv
             (bvor #xff2b01d5 user__9Vv58rJWt)
             #xffd5d5e9)
            (bvshl
             #xff8a9651
             (bvsdiv 'Box_3 #x00a12e6a)))))
         (bvor
          (bvadd
           (bvshl
            (bvneg
             (bvurem user__9Vv58rJWt #xffdd9b10))
            (bvashr
             (bvmul 'Box_1 'Box_0)
             (bvlshr 'Box_0 user__L_WC_omh_PtJGOPMl3UDM2fE)))
           (bvsdiv
            (bvor user__9Vv58rJWt #x0075e620)
            #xfe2b8ff2))
          (bvurem
           (bvnot
            (bvshl
             #x00000000
             (bvsdiv user__L_WC_omh_PtJGOPMl3UDM2fE #x00369396)))
           (bvudiv
            #xfb346d30
            (bvlshr
             (bvashr 'Box_3 #xfee01831)
             #xff1e1b48)))))
        (bvashr
         (bvashr
          (bvsrem
           (bvsdiv
            (bvadd
             #xc2cca010
             (bvneg
              (bvsdiv #xff329abe user__XMx4t__CwffO_4Uy_)))
            #xffffffff)
           #xfebaf9fd)
          #x00cd5bf9)
         (bvadd
          (bvadd
           #xff226e9e
           (bvashr
            (bvmul
             (bvmul #xfeffcb7c user__4nGak_7JBcgtPO_6xmkE)
             (bvlshr user__L_WC_omh_PtJGOPMl3UDM2fE 'Box_0))
            (bvlshr
             #xffffffff
             (bvmul user__XMx4t__CwffO_4Uy_ user__9Vv58rJWt))))
          (bvsdiv
           (bvmul
            (bvmul
             (bvsdiv user__XMx4t__CwffO_4Uy_ #xfffc00e1)
             (bvneg user__9Vv58rJWt))
            (bvadd
             #xff14174a
             (bvor user__XMx4t__CwffO_4Uy_ 'Box_1)))
           'Box_1)))))
      'Box_1))))))
```